### PR TITLE
feat(css/minifier) Follow the CSS spec more rigorously

### PR DESCRIPTION
**Pull request by ghostd**
**Description:**

Full refactor of the "calc" simplification to be more compliant with the spec

- [x] remove all FIXME
- [x] convert the units when possible

**BREAKING CHANGE:**

**Related issue (if exists):**

### CODE REVIEW

**Review by alexander-akait — COMMENTED** [review]
Please revert changes in `swc_html_parser` (in tests)?

File: crates/swc_css_minifier/tests/fixture.rs [hunk 1]
```diff
@@ -9,7 +9,8 @@ use swc_css_minifier::minify;
 use swc_css_parser::parse_file;
 use testing::NormalizedOutput;
 
-#[testing::fixture("tests/fixture/**/input.css")]
+// FIXME path
+#[testing::fixture("tests/fixture/compress-calc/**/input.css")]
```
Comment (alexander-akait) [inline crates/swc_css_minifier/tests/fixture.rs:13]: This need be fixed too :smile:

File: crates/swc_css_minifier/tests/fixture/compress-calc/convert-units/output.min.css [hunk 2]
```diff
@@ -1 +1 @@
...
+.class1,.class2{width:calc(1cm + 1px)}.class3{width:20q}.class4,.class5{width:calc(10px + 100.9q)}.class6{width:calc(10cm + 1px)}.class7{width:calc(1cm + 1px)}.class8{width:calc(10px + 1q)}.class9{width:calc(10cm + 1q)}.class10{width:calc(1cm + 1q)}.class11{width:calc(10in + 1q)}.class12{width:calc(10pt + 1q)}.class13{width:calc(10pc + 1q)}.class14{width:calc(10px + 1q)}.class15{width:calc(10cm + 1q)}.class16{width:calc(1cm + 1q)}.class17{width:calc(10in + 1q)}.class18{width:calc(10pt + 1q)}.class19{width:calc(10pc + 1q)}.class20{width:3unknown}.class21{width:calc(2px + 1unknown)}.class22{width:calc(1px + 2unknown)}
```
Comment (alexander-akait) [inline crates/swc_css_minifier/tests/fixture/compress-calc/convert-units/output.min.css:1]: Some changes are unexpected, for example - `.class9{width:calc(10cm + 1q)}`, can you look at this?

**Review by alexander-akait — COMMENTED** [review]
Most of changes are really good, big thank, look like we need transform some units to improve minification more - we can move them into own function and reuse it to get the same units

**Comment by alexander-akait** [comment]
@ghostd friendly ping

File: crates/swc_css_minifier/tests/fixture.rs (hunk 1 above)
Comment (ghostd) [reply crates/swc_css_minifier/tests/fixture.rs]: Fixed :)

File: crates/swc_css_minifier/tests/fixture/compress-calc/convert-units/output.min.css (hunk 2 above)
Comment (ghostd) [reply crates/swc_css_minifier/tests/fixture/compress-calc/convert-units/output.min.css]: Fixed

**Comment by ghostd** [comment]
We can do more optimizations, but i had rewritten all previous ones.

**Comment by alexander-akait** [comment]
@ghostd Thank you fro your work

> We can do more optimizations, but i had rewritten all previous ones.

Can you write them? Even more - can you put TODOs in code?

**Review by alexander-akait — CHANGES_REQUESTED** [review]
Let's remove any changes from [crates/swc_html_parser/tests/html5lib-tests](https://github.com/swc-project/swc/pull/6291/files#diff-e8984529f81ba2464c8bbdc677097cda803a75aaf1075f06ae745ba2292aa975)

File: crates/swc_css_minifier/Cargo.toml [hunk 3]
```diff
@@ -14,6 +14,7 @@ bench = false
 
 [dependencies]
 serde         = "1.0.118"
+string_cache  = "0.8.4"
```
Comment (alexander-akait) [inline crates/swc_css_minifier/Cargo.toml:17]: Please use `use swc_atoms::Atom;` instead `string_cache`

File: crates/swc_css_minifier/src/compressor/calc_sum.rs [hunk 4]
```diff
@@ -1,835 +1,807 @@
...
+    Function(Function),
+    OperatorNode(Box<CalcOperatorNode>),
+}
+
+fn get_precision(n: f64) -> u32 {
+    let n_as_str = n.to_string();
+    n_as_str
+        .find('.')
+        .map_or(0, |sep| (n_as_str.len() - sep) as u32 - 1)
+}
```
Comment (alexander-akait) [inline crates/swc_css_minifier/src/compressor/calc_sum.rs:36]: Why do not use `.fract()`?

File: crates/swc_css_minifier/tests/fixture/compress-calc/discard-zero/output.min.css [hunk 5]
```diff
@@ -1 +1 @@
-.class1{width:calc(100vw/2 - 6px)}.class2{width:500px}.class3{width:calc(0 - 10px)}.class4{width:calc(0 - 1px - 1em)}.class5{width:calc(0 - (100vw - 10px)/2)}.class6{width:calc(10px - 100vw)}.class7{width:calc(0px - (100vw - 10px)*2)}.class8{width:calc(-10px - 100vw)}.class9{width:calc(0px - var(--foo,4px)/2)}
+.class1{width:calc(-6px + 50vw)}.class2{width:500px}.class3{width:calc(0 - 10px)}.class4{width:calc(0 - 1em - 1px)}.class5{width:calc(0 + 5px - 50vw)}.class6{width:calc(10px - 100vw)}.class7{width:calc(20px - 200vw)}.class8{width:calc(-10px - 100vw)}.class9{width:calc(0px - .5*var(--foo,4px))}
```
Comment (alexander-akait) [inline crates/swc_css_minifier/tests/fixture/compress-calc/discard-zero/output.min.css:1]: `.class9{width:calc(0px - .5*var(--foo,4px))}` is valid compression, but `.class9{width:calc(0px - var(--foo,4px)/2)}` has less characters, can we avoid it in cases when characters are more (yes, it is not by spec, but in compresion it is normal to skip some cases in favor size)

Comment (alexander-akait) [reply crates/swc_css_minifier/tests/fixture/compress-calc/discard-zero/output.min.css:1]: Or we can put TODOs for future improvements

File: crates/swc_css_minifier/tests/fixture/packages/bootstrap_5_1_3/output.min.css [hunk 6]
```diff
@@ -1 +1 @@
...
...-y:auto}.modal-fullscreen-xxl-down .modal-footer{border-radius:0}}.tooltip{position:absolute;z-index:1080;display:block;margin:0;font-family:var(--bs-font-sans-serif);font-style:normal;font-weight:400;line-height:1.5;text-align:left;text-align:start;text-decoration:none;text-shadow:none;text-transform:none;letter-spacing:normal;word-break:normal;word-spacing:normal;white-space:normal;line-break:auto;font-size:.875rem;word-wrap:break-word;opacity:0}.tooltip.show{opacity:.9}.tooltip .tooltip-arrow{position:absolute;display:block;width:.8rem;height:.4rem}.tooltip .tooltip-arrow:before{position:absolute;content:"";border-color:transparent;border-style:solid}.bs-tooltip-top,.bs-tooltip-auto[data-popper-placement^=top]{padding:.4rem 0}.bs-tooltip-top .tooltip-arrow,.bs-tooltip-auto[data-popper-placement^=top] .tooltip-arrow{bottom:0}.bs-tooltip-top .tooltip-arrow:before,.bs-tooltip-auto[data-popper-placement^=top] .tooltip-arrow:before{top:-1px;border-width:.4rem .4rem 0;border-top-color:#000}.bs-tooltip-end,.bs-tooltip-auto[data-popper-placement^=right]{padding:0 .4rem}.bs-tooltip-end .tooltip-arrow,.bs-tooltip-auto[data-popper-placement^=right] .tooltip-arrow{left:0;width:.4rem;height:
```
Comment (alexander-akait) [inline crates/swc_css_minifier/tests/fixture/packages/bootstrap_5_1_3/output.min.css:1]: Here intresting case `calc(.75rem - 3px)` to `calc(-3px + .75rem)`, looks like for better compression we should change order and minus should be last... It is possible to do postprocessing after spec logic, so we will be align to spec and output better compression?

**Review by alexander-akait — COMMENTED** [review]
Great job, really :star: Let's fix edge cases and we can merge it

File: crates/swc_css_minifier/tests/fixture/packages/bootstrap_5_1_3/output.min.css (hunk 6 above)
Comment (ghostd) [reply crates/swc_css_minifier/tests/fixture/packages/bootstrap_5_1_3/output.min.css]: I fixed it.

File: crates/swc_css_minifier/tests/fixture/compress-calc/discard-zero/output.min.css (hunk 5 above)
Comment (ghostd) [reply crates/swc_css_minifier/tests/fixture/compress-calc/discard-zero/output.min.css]: I put a TODO on the `simplify_calc_operator_node_invert` function. I think it's a bit tricky to manage all cases.

File: crates/swc_css_minifier/src/compressor/calc_sum.rs (hunk 4 above)
Comment (ghostd) [reply crates/swc_css_minifier/src/compressor/calc_sum.rs:33]: `1.4.fract()` returns "0.33333333", but we need "1" instead to have the "right" precision. The rust `to_string()` function seems clever for this purpose.

File: crates/swc_css_minifier/Cargo.toml (hunk 3 above)
Comment (ghostd) [reply crates/swc_css_minifier/Cargo.toml]: Fixed

**Comment by ghostd** [comment]
I reverted the changes on the submodule "crates/swc_ecma_parser/tests/test262-parser".

**Review by alexander-akait — APPROVED** [review]
/cc @kdy1 We can merge it,there are some edge cases, but we have TODOs for them, so let's fix them in other PRs step by step

**Comment by ghostd** [comment]
@kdy1 done :)

**Review by kdy1 — APPROVED** [review]
Thank you!

---
---

---

swc-bump:
 - swc_css_minifier
